### PR TITLE
[IMPROVE] Open rooms quicker

### PR DESCRIPTION
**Pull request by rodrigok**
## Before
screen shot 2019-02-12 at 5 40 39 pm

## After
screen shot 2019-02-12 at 5 35 59 pm

### CODE REVIEW

File: packages/rocketchat-ui/client/views/app/room.js [hunk 1]
```diff
@@ -965,6 +929,54 @@ Template.room.onCreated(function() {
...
+		},
+	};
+
+	this.messages = new ReactiveVar([]);
+
+	const updateMessages = _.debounce(() => {
+		this.messages.set(ChatMessage.find(query, options).fetch());
+	}, 100);
+
+	ChatMessage.find(query).observe({
```
Comment (ggazzo) [inline packages/rocketchat-ui/client/views/app/room.js]: do you know if they cancel the observe after destroy the room? if not I think we gonna introduce a "event leak" here

Comment (rodrigok) [reply packages/rocketchat-ui/client/views/app/room.js]: Now it cancels :)

**Comment by ggazzo** [comment]
I made some tests and seems there are some bugs (mainly with scrolls)... We already made some solutions like that, and we never get the perfect solution =/

File: packages/rocketchat-ui/client/views/app/room.js [hunk 2]
```diff
@@ -229,43 +229,7 @@ Template.room.helpers({
...
-		}
-
-		const options = {
-			sort: {
-				ts: 1,
-			},
-		};
-
-		return ChatMessage.find(query, options);
+		return Template.instance().messages.get();
```
Comment (ggazzo) [inline packages/rocketchat-ui/client/views/app/room.js]: I made some tests here, and that helper just runs 1 time, is the cursor "ChatMessage.find(query, options)" slower than the other approach?

Comment (rodrigok) [reply packages/rocketchat-ui/client/views/app/room.js]: It is cuz it returns a cursor that triggers the render for every new record while here it's returning a reactive var that receives all data in a single operation.

**Comment by rodrigok** [comment]
@ggazzo fiz passar nos testes

**Comment by ggazzo** [comment]
hey @rodrigok its not just about the tests... please read that comment:
> I made some tests and seems there are some bugs (mainly with scrolls)... We already made some solutions like that, and we never get the perfect solution =/

I meant: 'I made some test with YOUR branch and the behavior its not good, scrolls are flickering'.

and more we cant use the aproach using debounce/throttle if we have sequential code before wainting the render.  

like: https://github.com/RocketChat/Rocket.Chat/blob/develop/packages/rocketchat-ui-utils/client/lib/RoomHistoryManager.js#L96
here we do calcs base on scroll position after the messages get inserted. you can insert a lot of problems doing thinks like that.

File: packages/rocketchat-ui/client/views/app/room.js [hunk 3]
```diff
@@ -965,9 +929,64 @@ Template.room.onCreated(function() {
...
+	const updateMessages = () => {
+		this.messages.set(ChatMessage.find(query, options).fetch());
+	};
+
+	const updateMessagesDebounced = _.debounce(() => {
+		this.messages.set(ChatMessage.find(query, options).fetch());
+	}, 100);
+
+	this.messageObserver = ChatMessage.find(query).observe({
+		added: process.env.TEST_MODE ? updateMessages : updateMessagesDebounced,
```
Comment (ggazzo) [inline packages/rocketchat-ui/client/views/app/room.js]: I dont like this approach (using `process.env.TEST_MODE`) but if we gonna use that... I would suggest few changes
```
const updateMessages = () => {
	this.messages.set(ChatMessage.find(query, options).fetch());
};
const updateMessagesDebounced = _.debounce(updateMessages, 100);
```
 more suggestions:
```

const _updateMessages = () => {
	this.messages.set(ChatMessage.find(query, options).fetch());
};
const updateMessages = process.env.TEST_MODE? _.debounce(_updateMessages, 100): _updateMessages ;

this.messageObserver = ChatMessage.find(query).observe({
  added: updateMessages,
  updated: updateMessages,
  removed: updateMessages,
})
```

**Comment by ggazzo** [comment]
@sampaiodiego, its not a delay its self... its a "two debounced upserts", but the user feels like a delay =/ , if you keep sending messages with a rate < debounce time, you will never render the messages.
> there is a huge delay between sending the message (hitting enter) and the message showing on the list.

**Comment by ggazzo** [comment]
@rodrigok if you pay attention here:
ezgif-2-77453934eea2
after load more messages, the scroll jumps to the "bottom". the reason I have explained above.

File: packages/rocketchat-ui/client/views/app/room.js (hunk 3 above)
Comment (rodrigok) [reply packages/rocketchat-ui/client/views/app/room.js]: It was removed

**Comment by rodrigok** [comment]
@ggazzo your explanation was wrong, the debounce has nothing to do with the scroll behavior or the delay rendering the message, both was caused by the full recalculation of the entire list caused by the usage of the ReactiveVar instead of the cursor. I moved back the the cursor cuz it wasn't the main improvement at the time.

The fail editing the message was caused by me using `updated` instead of `changed` in observeChanges.

I did some profilings and found the items that are most expensive being the find for permissions that happens several times per message render duo to the action buttons logic and the translate logic that execute finds agains the subscription collection (that may be big for certain users).

**Comment by ggazzo** [comment]
@rodrigok I gave up about discuss about use debounce or not (my explanation could not explain that specify bug, but prevent for many others), my final opinion (I dont think will change event with more chats): do not use debounce in some places, that case is one of them.

but I liked the memoize usage to avoid permissions recallcs (I think we should create a rocketchat helper/package to do that)

very smart the way you disabled the queries :)

I am testing now

**Pull request merged**